### ISSUE TITLE
catlas building enters infinite? loop

### ISSUE THREAD

**Issue by ctb** [spacegraphcats/spacegraphcats#57]
after
```
load-graph.py -M 16e9 -k 31 podar.ng SRR606249.pe.qc.fq.gz -n
python -m spacegraphcats.build_contracted_dbg -k 31 -M 16e9 -l podar.ng SRR606249.pe.qc.fq.gz -o podar
```

I ran:

```
python -m spacegraphcats.catlas podar/podar.gxt podar/podar.catlas podar/podar.domfile 3
```

which ran for about 24 hours before I killed it. After 8 hours, it stopped providing input at `Added edges between odd length optimal domination`, below.

```
Added edges between odd length optimal domination
Added edges between even length optimal domination
Catlas level 0 complete
Computing low degree orientation (step 1)
bucketed 7530335 of 7530335 nodes  
removed 7530335 of 7530335 nodes   
assigning to dominators
computing dominating edges
Added edges between odd length optimal domination
Added edges between even length optimal domination
Catlas level 1 complete
Computing low degree orientation (step 1)
bucketed 2995650 of 2995650 nodes
removed 2995650 of 2995650 nodes
assigning to dominators
computing dominating edges
Added edges between odd length optimal domination
Added edges between even length optimal domination
Catlas level 2 complete
Computing low degree orientation (step 1)
bucketed 1058059 of 1058059 nodes
removed 1058059 of 1058059 nodes
assigning to dominators
computing dominating edges
Added edges between odd length optimal domination
Added edges between even length optimal domination
Catlas level 3 complete
Computing low degree orientation (step 1)
bucketed 385257 of 385257 nodes
removed 385257 of 385257 nodes
assigning to dominators
computing dominating edges
Added edges between odd length optimal domination

Traceback (most recent call last):
  File "/usr/lib/python3.5/runpy.py", line 170, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/ubuntu/spacegraphcats/spacegraphcats/catlas.py", line 169, in <module>
    cProfile.run("main(args)")
  File "/usr/lib/python3.5/cProfile.py", line 16, in run
    return _pyprofile._Utils(Profile).run(statement, filename, sort)
  File "/usr/lib/python3.5/profile.py", line 55, in run
    prof.run(statement)
  File "/usr/lib/python3.5/cProfile.py", line 95, in run
    return self.runctx(cmd, dict, dict)
  File "/usr/lib/python3.5/cProfile.py", line 100, in runctx
    exec(cmd, globals, locals)
  File "<string>", line 1, in <module>
  File "/home/ubuntu/spacegraphcats/spacegraphcats/catlas.py", line 154, in main
    cat = CAtlas.build(G, r, args.domfile)
  File "/home/ubuntu/spacegraphcats/spacegraphcats/catlas.py", line 50, in build
    prev_nodes)
  File "/home/ubuntu/spacegraphcats/spacegraphcats/catlas.py", line 85, in _build_level
    domgraph, closest_dominators = domination_graph(graph, domset, radius)
  File "/home/ubuntu/spacegraphcats/spacegraphcats/rdomset.py", line 339, in domination_graph
    domgraph.in_neighbors(x, 1) - \
KeyboardInterrupt
```

**Comment by sgsmob** [comment]
Most likely this isn't an infinite loop.  In the section at which it stalled, we are looping over each edge in the current graph and drawing edges between the dominators at each of its endpoints.  When we have a dense area of the graph, the endpoints of each edge in the current graph may be dominated by lots of vertices, in which case the number of pairs of dominators to consider adding edges between is very large.  This can become a problem because we might be considering pairs of dominators that already have edges between them many times, causing very large amounts of unnecessary, repetitive computation.

When I wrote this section, I included some pruning that helped us skip over some of the redundant pairs because it was noticeably too slow without them.  Apparently there is not enough pruning going on.  I will think about how we can do this more efficiently.
